Add type annotations to torch._C._distributed_rpc module.

xuzhao9 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#46624 Add type annotations to torch._C._distributed_rpc module.**
* #46623 Add type annotations for torch._C._distributed_c10d module.
* #46622 Add Python declaration of torch._C and torch._C._autograd modules.

Differential Revision: [D24761656](https://our.internmc.facebook.com/intern/diff/D24761656)

xuzhao9 committed:
Add type annotations to torch._C._distributed_rpc module.

[ghstack-poisoned]

xuzhao9 committed:
Update on "Add type annotations to torch._C._distributed_rpc module."




[ghstack-poisoned]

xuzhao9 committed:
Update on "Add type annotations to torch._C._distributed_rpc module."




[ghstack-poisoned]

xuzhao9 committed:
Update on "Add type annotations to torch._C._distributed_rpc module."




[ghstack-poisoned]

xuzhao9 committed:
Update on "Add type annotations to torch._C._distributed_rpc module."




[ghstack-poisoned]

xuzhao9 committed:
Update on "Add type annotations to torch._C._distributed_rpc module."




[ghstack-poisoned]

walterddr commented on torch/_C/_distributed_rpc.pyi:
```diff
@@ -0,0 +1,191 @@
+from ._autograd import ProfilerConfig, ProfilerState
+from ._distributed_c10d import ProcessGroup, Store
+
+# This module is defined in torch/csrc/distributed/rpc/init.cpp
+
+# Defined in torch/csrc/distributed/rpc/init.cpp
```
nit: duplicate comment

walterddr commented on torch/_C/_distributed_rpc.pyi:
```diff
@@ -0,0 +1,191 @@
+_UNSET_RPC_TIMEOUT: float
+_DEFAULT_RPC_TIMEOUT_SEC: float
+
+class RpcBackendOptions:
```
general question - some of the definitions here mismatched with original cpp. to name a few:
1 `rpcTimeoutSeconds` in this class is not defined 
2. `worker_id` for the init method in `WorkerInfo` is named as `id`

any reason of the discrepancy here?

xuzhao9 replied:
This is probably because the cpp file interface changes comparing with the time when I write the pyi file. Therefore, I need to update the python interface file to reflect the change.

walterddr reviewed (commented):
looks good overall. thanks for breaking the bigger PR into the small ones!

malfet reviewed (changes requested):
@xuzhao9 please rebase your change on top of master and fix compilation errors

xuzhao9 committed:
Update on "Add type annotations to torch._C._distributed_rpc module."




[ghstack-poisoned]

xuzhao9 committed:
Update on "Add type annotations to torch._C._distributed_rpc module."




[ghstack-poisoned]

xuzhao9 committed:
Update on "Add type annotations to torch._C._distributed_rpc module."




[ghstack-poisoned]

xuzhao9 committed:
Update on "Add type annotations to torch._C._distributed_rpc module."




[ghstack-poisoned]

xuzhao9 committed:
Update on "Add type annotations to torch._C._distributed_rpc module."




[ghstack-poisoned]

xuzhao9 committed:
Update on "Add type annotations to torch._C._distributed_rpc module."




[ghstack-poisoned]

xuzhao9 committed:
Update on "Add type annotations to torch._C._distributed_rpc module."




[ghstack-poisoned]

xuzhao9 committed:
Update on "Add type annotations to torch._C._distributed_rpc module."




[ghstack-poisoned]

xuzhao9 committed:
Update on "Add type annotations to torch._C._distributed_rpc module."




[ghstack-poisoned]

xuzhao9 committed:
Update on "Add type annotations to torch._C._distributed_rpc module."




[ghstack-poisoned]

xuzhao9 committed:
Update on "Add type annotations to torch._C._distributed_rpc module."




[ghstack-poisoned]

xuzhao9 committed:
Update on "Add type annotations to torch._C._distributed_rpc module."




[ghstack-poisoned]

xuzhao9 committed:
Update on "Add type annotations to torch._C._distributed_rpc module."




[ghstack-poisoned]

xuzhao9 committed:
Update on "Add type annotations to torch._C._distributed_rpc module."




[ghstack-poisoned]

xuzhao9 committed:
Update on "Add type annotations to torch._C._distributed_rpc module."




[ghstack-poisoned]

xuzhao9 committed:
Update on "Add type annotations to torch._C._distributed_rpc module."




[ghstack-poisoned]

xuzhao9 committed:
Update on "Add type annotations to torch._C._distributed_rpc module."




[ghstack-poisoned]

xuzhao9 committed:
Update on "Add type annotations to torch._C._distributed_rpc module."




[ghstack-poisoned]

xuzhao9 committed:
Update on "Add type annotations to torch._C._distributed_rpc module."




[ghstack-poisoned]

xuzhao9 committed:
Update on "Add type annotations to torch._C._distributed_rpc module."




[ghstack-poisoned]

xuzhao9 committed:
Update on "Add type annotations to torch._C._distributed_rpc module."




[ghstack-poisoned]

malfet reviewed (approved):
Looks good to me, but can you please modify `mypy.ini` to add `torch.distributed.rpc` to the set of modules where type annotation is enforced.

xuzhao9 committed:
Update on "Add type annotations to torch._C._distributed_rpc module."




[ghstack-poisoned]

xuzhao9 commented on torch/distributed/rpc/internal.py:
```diff
@@ -224,8 +227,8 @@ def _start_record_function(exec_type, func_name, current_worker_name, dest_worke
     )
-    rf = torch.autograd._RecordFunction()
-    torch.autograd._run_before_callbacks(rf, profile_key)
+    rf = torch.autograd._RecordFunction()  # type: ignore[attr-defined]
```
Ignoring this error because I don't find anywhere "_RecordFunction()" and "_run_before_callbacks()" are defined.

xuzhao9 commented on torch/distributed/rpc/api.py:
```diff
@@ -379,16 +379,18 @@ def _rref_typeof_on_user(rref):
     # must be a (non-strict) subclass of the metaclasses of all its bases
-    class RRefMeta(PyRRef.__class__, GenericWithOneTypeVar.__class__):
+    # Mypy doesn't understand __class__ (mypy bug #4177)
+    class RRefMeta(PyRRef.__class__, GenericWithOneTypeVar.__class__):  # type: ignore
```
Ignore all type errors because this line (and line 393) will cause multiple type errors.

xuzhao9 committed:
Update on "Add type annotations to torch._C._distributed_rpc module."




[ghstack-poisoned]

xuzhao9 committed:
Update on "Add type annotations to torch._C._distributed_rpc module."




[ghstack-poisoned]

xuzhao9 commented on torch/distributed/rpc/api.py:
```diff
@@ -379,16 +379,18 @@ def _rref_typeof_on_user(rref):
 try:
     # Combine the implementation class and the type class.
-    class RRef(PyRRef, GenericWithOneTypeVar):
+    class RRef(PyRRef, Generic[T]):
```
mypy complains " Variable "torch.distributed.rpc.api.GenericWithOneTypeVar" is not valid as a type  [valid-type]" if use GenericWithOneTypeVar here.

xuzhao9 committed:
Update on "Add type annotations to torch._C._distributed_rpc module."




[ghstack-poisoned]

xuzhao9 commented on torch/csrc/distributed/rpc/init.cpp:
```diff
@@ -114,6 +120,20 @@ PyObject* rpc_init(PyObject* _unused, PyObject* noargs) {
               "join", &RpcAgent::join, py::call_guard<py::gil_scoped_release>())
           .def(
               "sync", &RpcAgent::sync, py::call_guard<py::gil_scoped_release>())
+          .def(
```
Adding three definitions because they are used in internal.py and api.py.

xuzhao9 commented:
> Looks good to me, but can you please modify `mypy.ini` to add `torch.distributed.rpc` to the set of modules where type annotation is enforced.

Thanks for the review. I have modified `mypy.ini` to enable the type checking related to this stack of changes. Note I have to introduce some changes to eliminate the errors. Please see my follow-up comments and let me know if you have any feedbacks.

xuzhao9 committed:
Update on "Add type annotations to torch._C._distributed_rpc module."




[ghstack-poisoned]

xuzhao9 committed:
Update on "Add type annotations to torch._C._distributed_rpc module."




[ghstack-poisoned]

xuzhao9 committed:
Update on "Add type annotations to torch._C._distributed_rpc module."


Differential Revision: [D24761656](https://our.internmc.facebook.com/intern/diff/D24761656)

[ghstack-poisoned]

xuzhao9 committed:
Update on "Add type annotations to torch._C._distributed_rpc module."


Differential Revision: [D24761656](https://our.internmc.facebook.com/intern/diff/D24761656)

[ghstack-poisoned]
